Source/WTF/wtf/AliasedRefPtr.h

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+namespace WTF {
+
+template <typename VisibleType>
+class AliasedRefPtr {
```

Comment:
I have trouble understanding what this class is for and why we can't use RefPtr, even after reading the patch / commit log. Please clarify why we need yet another smart pointer type in WTF.

Reply:
This was primarily intended to deal with SerialFunctionDispatcher.

SerialFunctionDispatcher isn't a ref counted class.
However, there are SerialFunctionDispatcher that are ref-counted: WorkQueue, WorkerThread, RunLoop in particular.

This allows to generically store a SerialFunctionDispatcher, so long as the derived class has ref/deref method, without having the user of the class know which actually it's being given, which would require to use std::variants or template the class.

A ManagedSerialFunctionDispatcher is an actual type, it's no longer a templated type, so you design your method to take a ManagedSerialFunctionDispatcher and it can do so without knowing the derived class type.

I don't believe this can be done with RefPtr template parameter.

I consider the alternative of having this template class in FunctionDispatcher.h but then we have issue with headers.

The other alternative was to make SerialFunctionDispatcher ref-counted, but that is opening a can of worms as there are multiple instances of SerialFunctionDispatcher that aren't ref-counted.

AliasedRefPtr on its own isn't particularly needed, but I felt it would be useful to have generically.

Reply:
I amended the commit log. Hopefully it's clearer now

Reply:
Following discussion on slack, I will add a new RefCountedSerialFunctionDispatcher, which not only guarantees that it is ref-counted, but also that its API contract lifetime is the same as the object lifetime.
For now the only two SerialFunctionDispatcher guaranteeing this are WorkQueue and RunLoop.